File: lib/domain/logging/stream-ticker.ts
import type { MealItem } from '@/lib/core/types/meal';

/**
 * One frame of the ticker; a new `key` flips the text in place.
 *
 * `phase` frames are the app talking about itself and carry the cycling action
 * verbs; `item`/`macros` frames carry the meal's own words (serif italic), with
 * `detail` as the resolved kcal.
 */
export interface StreamTickerFrame {
  key: string;
  kind: 'phase' | 'item' | 'macros';
  text: string;
  detail?: string;
  /**
   * The pipeline stage this frame was derived in, which is what the verbs are
   * keyed on. Carried by dish frames too: between dishes the line falls back to
   * the stage's verbs, and without it the first name detected during
   * `decomposing` would hold the line for the rest of the run.
   *
   * Null for frames with no stage behind them (saving), which show `text`.
   */
  stage: string | null;
}

/** What the ticker reads off a live analysis, however it was carried there. */
export interface StreamTickerSource {
  isAnalyzing: boolean;
  status: string;
  /** Dish names as they are detected (decomposing), newest last. */
  items: string[];
  /** Dishes with resolved macros (estimating), newest last. */
  completedItems: MealItem[];
}

/**
 * The ONE line the ticker shows, derived from the stream rather than
 * accumulated.
 *
 * The pipeline emits names (decomposing) before macros (estimating), so
 * precedence is: latest resolved item > latest named item > the stage itself.
 * Assembling is excluded from the dish branches — by then the dishes are old
 * news and the line should say what is still happening.
 *
 * Shared by the dashboard's input bar and the logging feed so the two cannot
 * tell different stories about the same run.
 */
export function deriveStreamTicker(
  stream: StreamTickerSource
): StreamTickerFrame | null {
  if (!stream.isAnalyzing) return null;
  if (stream.status !== 'assembling') {
    const done = stream.completedItems.at(-1);
    if (done) {
      return {
        key: `done-${done.id}`,
        kind: 'macros',
        text: done.name,
        detail: `${Math.round(done.macros.calories)} kcal`,
        stage: stream.status,
      };
    }
    const name = stream.items.at(-1);
    if (name) {
      return {
        key: `name-${stream.items.length}`,
        kind: 'item',
        text: `${name}…`,
        stage: stream.status,
      };
    }
  }
  return {
    key: `phase-${stream.status}`,
    kind: 'phase',
    text: '',
    stage: stream.status,
  };
}

File: apps/mobile-flutter/lib/features/logging/logic/feed/stream_ticker.dart
/// The one line the streaming surface shows, derived fresh from the analysis
/// state — the Dart port of the web dashboard/circle meal bar's ticker
/// (`hooks/dashboard/use-dashboard-meal-log.tsx`).
///
/// Frames carry l10n KEYS, not resolved text, so the derivation is a pure
/// function testable without an `easy_localization` bootstrap.
library;

import '../../../../models/logging/streaming.dart';
import '../../data/stream_analysis_controller.dart';

/// Statuses with a phase label of their own; everything else falls back to
/// "Analyzing…" (the web `getStreamingPhaseLabel` default).
String phaseLabelKey(StreamStatus status) => switch (status) {
  StreamStatus.connecting => 'logging.streaming.connecting',
  StreamStatus.decomposing => 'logging.streaming.decomposing',
  StreamStatus.matching => 'logging.streaming.matching',
  StreamStatus.estimating => 'logging.streaming.estimating',
  StreamStatus.assembling => 'logging.streaming.assembling',
  _ => 'logging.streaming.analyzing',
};

/// One frame of the ticker. A new [key] flips the line; an unchanged key must
/// leave it alone, so keys are derived from the DATA and never from a counter
/// that would churn on every rebuild.
sealed class StreamTickerFrame {
  const StreamTickerFrame();

  String get key;
}

/// A pipeline stage — quiet muted sans.
final class PhaseFrame extends StreamTickerFrame {
  const PhaseFrame(this.status);

  final StreamStatus status;

  String get labelKey => phaseLabelKey(status);

  @override
  String get key => 'phase-${status.name}';
}

/// A dish the pipeline just named but hasn't priced yet — the meal's own words,
/// rendered with a trailing ellipsis.
final class ItemFrame extends StreamTickerFrame {
  const ItemFrame({required this.name, required this.count});

  final String name;

  /// How many names had arrived when this frame was built. Part of the key so
  /// two dishes with the same name still flip the line.
  final int count;

  @override
  String get key => 'name-$count';
}

/// A dish that just resolved — its name plus the calories it landed on.
final class MacrosFrame extends StreamTickerFrame {
  const MacrosFrame({
    required this.id,
    required this.name,
    required this.calories,
  });

  final String id;
  final String name;
  final int calories;

  @override
  String get key => 'done-$id';
}

/// The current frame, or null when nothing is in flight.
///
/// Precedence is the web's, 1:1: assembling label > latest resolved item >
/// latest named item > stage label. The `assembling` carve-out is deliberate —
/// once the pipeline is assembling, the line stops advertising dishes and says
/// what it is doing.
///
/// The web's first branch (`isSaving`) has NO mobile counterpart: the mobile
/// flow ends at a user-confirmed reveal card, not an auto-save, so there is no
/// saving phase to announce.
StreamTickerFrame? deriveStreamTicker(StreamAnalysisState stream) {
  if (!stream.isAnalyzing) return null;

  if (stream.status != StreamStatus.assembling) {
    if (stream.completedItems.isNotEmpty) {
      final done = stream.completedItems.last;
      return MacrosFrame(
        id: done.id,
        name: done.name,
        calories: done.macros.calories.round(),
      );
    }
    if (stream.items.isNotEmpty) {
      return ItemFrame(name: stream.items.last, count: stream.items.length);
    }
  }

  return PhaseFrame(stream.status);
}
